pkg/compiler/namespace.go

```diff
@@ -0,0 +1,97 @@
+package compiler
+
+import (
+	"github.com/MontFerret/ferret/pkg/runtime/core"
+	"github.com/pkg/errors"
+	"regexp"
+	"strings"
+)
+
+var fnNameValidation = regexp.MustCompile("^[a-zA-Z]+[a-zA-Z0-9_]*(::[a-zA-Z]+[a-zA-Z0-9_]*)*$")
+
+const emptyNS = ""
+const separator = "::"
+
+type NamespaceBuilder struct {
+	funcs core.Functions
+	name  string
+}
+
+func newRootNamespace() *NamespaceBuilder {
+	ns := new(NamespaceBuilder)
+	ns.funcs = make(core.Functions)
+
+	return ns
+}
+
+func newNamespace(funcs core.Functions, name string) *NamespaceBuilder {
+	return &NamespaceBuilder{funcs, strings.ToUpper(name)}
+}
+
+func (rs *NamespaceBuilder) Namespace(name string) core.Namespace {
+	return newNamespace(rs.funcs, rs.makeFullName(name))
+}
+
+func (rs *NamespaceBuilder) RegisterFunction(name string, fun core.Function) error {
+	nsName := rs.makeFullName(name)
+	_, exists := rs.funcs[nsName]
+
+	if exists {
+		return errors.Errorf("function already exists: %s", name)
```

Comment:
Why not just return the nil?

Reply:
nil of what?

Reply:
of error

Reply:
why?

Reply:
because function is already registered and it doesn't look like an error

Reply:
This is the point of the error. You get an error if you try to register a function with already used name. (Name collisions).

Reply:
Error is better. You'he right.